### DISCUSSION TITLE
Long-Lived Reference Types in Views

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#2497] [Q&A]
Some views in SwiftUI require long-lived reference types such as VideoPlayer which expect to receive an AVPlayer. While I can retain these objects locally within views typically using something like @StateObject my goal is to have the same object accessible via a dependency so that I can play with it (for various reasons: like stop/play video) as well. I'm curious if anyone has come across or can provide examples of how to address this in The Composable Architecture. Any insights or examples would be greatly appreciated.

**Comment by rzolin** [comment]
Have you tried to have a singleton wrapped as a dependency? I use dependency to access a framework with a bunch of services and states in singleton(s). So if your singleton can store references to instance(s) of VideoPlayer and such, you can wrap it in dependency and access via it.

Here is abbreviated version of how I use it:

```swift

public protocol ATConfigDependency {
    var primaryUserAccount: ATUserAccount? { get }
}

public struct ATConfigClient_Live: ATConfigDependency {
    
    public var primaryUserAccount: ATUserAccount? {
        return ATSDK.config.primaryUserId
    }

}
```

Note: I usually have a separate method to update such objects, something like `updateUserAccount(account: ATUserAccount)`

And later on:
```swift

@Dependency(\.config) private var config

{
    print(config.primaryUserAccount)
}

```
